src/openrct2-ui/windows/TopToolbar.cpp

```diff
@@ -1906,12 +1907,26 @@ static void window_top_toolbar_scenery_tool_down(int16_t x, int16_t y, rct_windo
         }
         case SCENERY_TYPE_BANNER:
         {
-            int32_t flags = (parameter_1 & 0xFF00) | GAME_COMMAND_FLAG_APPLY;
-
-            gGameCommandErrorTitle = STR_CANT_POSITION_THIS_HERE;
-            game_command_callback = game_command_callback_place_banner;
-            game_do_command(
-                gridX, flags, gridY, parameter_2, GAME_COMMAND_PLACE_BANNER, parameter_3, gWindowSceneryPrimaryColour);
+            uint8_t direction = (parameter_2 >> 8) & 0xFF;
+            int32_t z = (parameter_2 & 0xFF) * 16;
+            CoordsXYZD loc{ gridX, gridY, z, direction };
+            auto primaryColour = gWindowSceneryPrimaryColour;
+            auto bannerType = (parameter_1 & 0xFF00) >> 8;
+            auto bannerIndex = create_new_banner(0);
```

Comment:
I'm a little worried about what if two clients are placing banners at the same time, this would collide

Reply:
It would fail the check in the game action if two tried at the same tick.

Reply:
Not sure how I feel about that if I were to fight with someone placing banners at the same time I am, either I or the other person would basically get randomly some error, I mean its not that likely to happen but if two people are currently mass placing banners they will notice. One way to compensate this a little bit could by trying it a few times if the callback says via error message that there is a collision of banner index and then just use the next free one and try again, but then it would better if ghost sprites were handled differently, best not stored in the sprite list at all rather having a vector of sprites that are supposed to be ghost pieces.

Reply:
This is the same principle used for all of the banner index allocations.